libsupport/src/ArrowInterchange.cpp

```diff
@@ -70,3 +196,243 @@ katana::EmptyChunkedArray(
   std::vector<std::shared_ptr<arrow::Array>> chunks{maybe_array.ValueOrDie()};
   return std::make_shared<arrow::ChunkedArray>(chunks);
 }
+
+template <typename BuilderType, typename ScalarType>
+katana::Result<std::shared_ptr<arrow::Array>>
+BasicToArray(
```

Comment:
```c++
struct BasicToArrayVisitor {
  using ReturnType = std::shared_ptr<arrow::Array>;
  using ResultType = katana::Result<ReturnType>;

  template <typename ArrowType, typename BuilderType>
  arrow::enable_if_t<
      arrow::is_number_type<ArrowType>::value ||
          arrow::is_boolean_type<ArrowType>::value ||
          arrow::is_temporal_type<ArrowType>::value,
      ResultType>
  Call(BuilderType* builder) {
    // same thing as BasicToArray, uses scalar.value
  }

  template <typename ArrowType, typename BuilderType>
  arrow::enable_if_string_like<ArrowType, ResultType>
  Call(BuilderType* builder) {
    // same thing as BasicToArray, uses scalar.value->ToString()
  }
  
  // other functions for List and Map, ok to just fail for now probably
  // arrow:: enable_if_list_type
  // arrow::enable_if_struct (Map)

  const std::vector<std::shared_ptr<arrow::Scalar>>& scalars;
};

// use arrow::MakeBuilder to pre-create the Builder
katana::Result<std::shared_ptr<arrow::Array>>
ArrayFromScalars(const std::vector<std::shared_ptr<arrow::Scalar>>& scalars, const shared_ptr<arrow::Datatype>& type) {
  std::unique_ptr<arrow::ArrayBuilder> builder;
  if (auto st = arrow::MakeBuilder(arrow::default_memory_pool(), type, &builder); !st.ok()) {...}

  BasicToArrayVisitor visitor{
    .scalars = scalars,
  };
  return VisitArrow(builder, visitor);
}
```

Reply:
This is deep @danielmawhirter, thank you.  I'm trying to use this code as a starting place and I'm making progress, but I'm a little baffled at the `.scalars = scalars` initialization and so is my compiler.  It says "scalars" doesn't refer to any field in type BasicToArrayVisitor.  I can't just add `const std::vector<std::shared_ptr<arrow::Scalar>>& scalars;` because that gets me back to having to cast the scalar to get its `value`.

Reply:
Also, where does `visit::Arrow` live?

Reply:
I probably got that syntax wrong, it's just trying to make an instance of that struct with the reference member initialized (might just want a constructor).
`VisitArrow` currently lives [in libquery](https://github.com/KatanaGraph/katana-enterprise/blob/master/libquery/include/katana/query/ArrowVisitor.h) but I consider it ready to move into a more central location, in the open repo (probably leaving the big `AppendScalarToBuilder` definition in libquery). The definitions are actually fairly concise, all things considered

Reply:
katana-enterprise/libquery/include/katana/query/ArrowVisitor.h has no dependencies in libquery, in preparation for this migration

Reply:
I know this is a complicated snippet and I don't mean to force you to specify everything.  But by "reference member initialized" to you mean `const std::vector<std::shared_ptr<arrow::Scalar>>&`?  Because that is my scalars input.  But if I get a reference to that vector, I am left with a `std::shared_ptr<arrow::Scalar>` which I have to cast to get the value.

I want to land this PR soon.  Can I just move `VisitArrow `into `ArrowInterchange`?

Reply:
Oh, crossed messages.  So I will make it so `ArrowInterchange.cpp` in `libsupport `can include `ArrowVisitor.h`

Reply:
Oh, I can't do that because `ArrowVisitor.h` is in enterprise and libsupport is in open.

Reply:
Let's add a constructor and change the initialization:
```c++
BasicToArrayVisitor::BasicToArrayVisitor (const std::vector<std::shared_ptr<arrow::Scalar>>& input) : scalars(input) {}

BasicToArrayVisitor visitor(scalars);
```

Then yes, inside each `Call` function you'll need code like:
```c++
using ScalarType = typename arrow::TypeTraits<ArrowType>::ScalarType;
const std::shared_ptr<arrow::Scalar>& untyped_scalar= scalars[0];
const ScalarType* typed_scalar = static_cast<ScalarType*>(untyped_scalar.get());
builder->UnsafeAppend(typed_scalar->value);
```

And yes, everything in ArrowVisitor.h, other than `AppendScalarToBuilder`, should move to ArrowInterchange

Reply:
I think this is the crux of the biscuit.

```
using ScalarType = typename arrow::TypeTraits<ArrowType>::ScalarType; const std::shared_ptr<arrow::Scalar>& 
 const ScalarType* typed_scalar = static_cast<ScalarType*>(untyped_scalar.get());
```

We are casting the untyped scalar, but casting it to what?  Is ScalarType resolved at compile time?

Reply:
The only thing resolved at runtime is "What is the type of this Builder? Call into that template environment". Once there, we know the ArrowType, BuilderType, and ScalarType at compile time. Since this is `static_cast` there's no runtime overhead (because it's an unchecked cast).
To put it another way, the template `ArrowType = arrow::Int64Type` implies `BuilderType = arrow::Int64Builder` and `ScalarType = arrow::Int64Scalar`, which are all resolved at compile time.